lib/SyTest/Federation/Datastore.pm

```diff
@@ -0,0 +1,250 @@
+package SyTest::Federation::Datastore;
+
+use strict;
+use warnings;
+
+use Carp;
+
+use Protocol::Matrix qw( sign_event_json );
+
+use List::MoreUtils qw( uniq );
+use Time::HiRes qw( time );
+
+sub new
+{
+   my $class = shift;
+   my %args = @_;
+
+   return bless {
+      %args,
+      keys => {},
+
+      next_event_id => 0,
+      next_room_id => 0,
+   }, $class;
+}
+
+=head2 server_name
+
+   $name = $store->server_name
+
+Returns the federation name of the server
+
+=cut
+
+sub server_name { $_[0]->{server_name} }
+
+=head2 key_id
+
+   $id = $store->key_id
+
+Returns the key ID of the signing key the server is currently using
+
+=cut
+
+sub key_id { $_[0]->{key_id} }
+
+=head2 public_key
+
+   $key = $store->public_key
+
+=head2 secret_key
+
+   $key = $store->secret_key
+
+Return the public or secret halves of the signing key the server is currently
+using
+
+=cut
+
+sub public_key { $_[0]->{public_key} }
+sub secret_key { $_[0]->{secret_key} }
+
+=head2 sign_event
+
+   $store->sign_event( $event )
+
+Applies the event signing algorithm to the given event, adding the result to
+the C<signatures> key.
+
+=cut
+
+sub sign_event
+{
+   my $self = shift;
+   my ( $event ) = @_;
+
+   sign_event_json( $event,
+      secret_key => $self->secret_key,
+      origin     => $self->server_name,
+      key_id     => $self->key_id,
+   );
+}
+
+=head2 get_key
+
+   $key = $store->get_key( server_name => $name, key_id => $id )
+
+=head2 put_key
+
+   $store->put_key( server_name => $name, key_id => $id, key => $key )
+
+Accessor and mutator for federation key storage
+
+=cut
+
+sub get_key
```

Comment:
It relates to our local cache of remote server (federation) signing keys. I'll add to docs.